File: scripts/compute_portfolio_opt.py
"""
Portfolio Optimization — Efficient Frontier + Nash Equilibrium
==============================================================
Combines three frameworks to identify superior projects:

  1. EVM (CPI/SPI)       — performance-based risk/return scores
  2. Markowitz Frontier  — Pareto-optimal projects in (risk, return) space
  3. Nash Equilibrium    — projects where all stakeholder roles are satisfied

Output: portfolio_recommendation table populated with tier labels
        (SELECT / CONSIDER / MONITOR / DEFER) and combined scores.

Run:
  DATABASE_URL=... python3.11 scripts/compute_portfolio_opt.py
"""

import os
import sys
import warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
import psycopg2
import psycopg2.extras

# ── Database connection ────────────────────────────────────────────────────────
DB_URL = os.environ.get(
    "DATABASE_URL",
    "postgresql://claudegabriel@localhost:5433/nme_portfolio"
)

def get_conn():
    return psycopg2.connect(DB_URL)

# ── 1. Pull project portfolio base data ───────────────────────────────────────
PORTFOLIO_SQL = """
SELECT
    project_id,
    project_code,
    project_name,
    project_status,
    trial_id,
    nct_number,
    trial_phase,
    nme_id,
    nme_name,
    nme_code,
    therapeutic_area,
    bac,
    pv,
    ev,
    ac,
    task_completion_pct,
    spi,
    cpi,
    return_score,
    risk_score,
    ps_payoff,
    mm_payoff,
    ra_payoff
FROM v_project_portfolio_base
"""

# ── 2. Pareto frontier (non-dominated set) ────────────────────────────────────
def compute_pareto_frontier(df: pd.DataFrame) -> pd.Series:
    """
    A project is on the efficient frontier if no other project dominates it.
    Project A dominates B if: A.return_score >= B.return_score
                          AND A.risk_score   <= B.risk_score
                          AND at least one inequality is strict.
    Returns a boolean Series indexed like df.
    """
    returns = df["return_score"].values.astype(float)
    risks   = df["risk_score"].values.astype(float)
    n = len(returns)
    is_frontier = np.ones(n, dtype=bool)

    for i in range(n):
        for j in range(n):
            if i == j:
                continue
            # j dominates i if j is at least as good in both dims AND strictly better in one
            if (returns[j] >= returns[i] and risks[j] <= risks[i] and
                    (returns[j] > returns[i] or risks[j] < risks[i])):
                is_frontier[i] = False
                break

    return pd.Series(is_frontier, index=df.index)

# ── 3. Nash equilibrium preference ────────────────────────────────────────────
def compute_nash_preferred(df: pd.DataFrame) -> pd.Series:
    """
    A project is 'Nash preferred' when all three stakeholder roles have an
    above-average payoff on that project — meaning no role would benefit
    from unilaterally moving to a different project.

    Payoffs are task_completion_pct (0–100) per role per project.
    """
    mean_ps = df["ps_payoff"].mean()
    mean_mm = df["mm_payoff"].mean()
    mean_ra = df["ra_payoff"].mean()

    nash = (
        (df["ps_payoff"] >= mean_ps) &
        (df["mm_payoff"] >= mean_mm) &
        (df["ra_payoff"] >= mean_ra)
    )
    return nash

# ── 4. Combined score ─────────────────────────────────────────────────────────
def compute_combined_score(df: pd.DataFrame) -> pd.Series:
    """
    Blended score (0–1):
      40% → return_score          (EVM performance)
      40% → 1 − risk_score        (lower risk is better)
      20% → mean payoff / 100     (stakeholder alignment, normalised)
    """
    mean_payoff = (df["ps_payoff"] + df["mm_payoff"] + df["ra_payoff"]) / 3.0 / 100.0
    score = (
        0.40 * df["return_score"].clip(0, 1) +
        0.40 * (1 - df["risk_score"].clip(0, 1)) +
        0.20 * mean_payoff.clip(0, 1)
    )
    return score.round(4)

# ── 5. Recommendation tier ────────────────────────────────────────────────────
def assign_tier(row: pd.Series, median_score: float) -> str:
    if row["is_frontier"] and row["nash_preferred"]:
        return "SELECT"
    if row["is_frontier"] or row["nash_preferred"]:
        return "CONSIDER"
    if row["combined_score"] >= median_score:
        return "MONITOR"
    return "DEFER"

# ── 6. Main ───────────────────────────────────────────────────────────────────
def main():
    print("Connecting to database…")
    conn = get_conn()
    cur  = conn.cursor(cursor_factory=psycopg2.extras.RealDictCursor)

    cur.execute(PORTFOLIO_SQL)
    rows = cur.fetchall()
    df = pd.DataFrame(rows)
    print(f"  Loaded {len(df)} projects from v_project_portfolio_base")

    if df.empty:
        print("No projects found — nothing to compute.")
        conn.close()
        return

    # Cast Decimal/object columns to float
    for col in ["bac", "pv", "ev", "ac", "task_completion_pct",
                "spi", "cpi", "return_score", "risk_score",
                "ps_payoff", "mm_payoff", "ra_payoff"]:
        df[col] = df[col].apply(lambda x: float(x) if x is not None else 0.0)

    # ── Compute framework metrics ──────────────────────────────────────────────
    df["is_frontier"]    = compute_pareto_frontier(df)
    df["nash_preferred"] = compute_nash_preferred(df)
    df["combined_score"] = compute_combined_score(df)

    median_score = float(df["combined_score"].median())
    df["recommendation"] = df.apply(assign_tier, axis=1, median_score=median_score)

    # ── Summary ────────────────────────────────────────────────────────────────
    tier_counts = df["recommendation"].value_counts()
    frontier_n  = df["is_frontier"].sum()
    nash_n      = df["nash_preferred"].sum()
    print(f"\n  Efficient frontier: {frontier_n} projects")
    print(f"  Nash preferred:     {nash_n} projects")
    print(f"  Tier distribution  —  SELECT: {tier_counts.get('SELECT', 0)}  "
          f"CONSIDER: {tier_counts.get('CONSIDER', 0)}  "
          f"MONITOR: {tier_counts.get('MONITOR', 0)}  "
          f"DEFER: {tier_counts.get('DEFER', 0)}")

    # ── Upsert to DB ───────────────────────────────────────────────────────────
    print("\nWriting recommendations to database…")
    upsert_sql = """
    INSERT INTO portfolio_recommendation
        (project_id, return_score, risk_score, spi, cpi,
         is_frontier, nash_preferred,
         ps_payoff, mm_payoff, ra_payoff,
         recommendation, combined_score, computed_at)
    VALUES
        (%(project_id)s, %(return_score)s, %(risk_score)s, %(spi)s, %(cpi)s,
         %(is_frontier)s, %(nash_preferred)s,
         %(ps_payoff)s, %(mm_payoff)s, %(ra_payoff)s,
         %(recommendation)s, %(combined_score)s, NOW())
    ON CONFLICT (project_id) DO UPDATE SET
        return_score   = EXCLUDED.return_score,
        risk_score     = EXCLUDED.risk_score,
        spi            = EXCLUDED.spi,
        cpi            = EXCLUDED.cpi,
        is_frontier    = EXCLUDED.is_frontier,
        nash_preferred = EXCLUDED.nash_preferred,
        ps_payoff      = EXCLUDED.ps_payoff,
        mm_payoff      = EXCLUDED.mm_payoff,
        ra_payoff      = EXCLUDED.ra_payoff,
        recommendation = EXCLUDED.recommendation,
        combined_score = EXCLUDED.combined_score,
        computed_at    = EXCLUDED.computed_at
    """

    records = df[[
        "project_id", "return_score", "risk_score", "spi", "cpi",
        "is_frontier", "nash_preferred",
        "ps_payoff", "mm_payoff", "ra_payoff",
        "recommendation", "combined_score",
    ]].to_dict("records")

    # Convert numpy booleans → Python booleans for psycopg2
    for rec in records:
        rec["is_frontier"]    = bool(rec["is_frontier"])
        rec["nash_preferred"] = bool(rec["nash_preferred"])
        rec["spi"]            = float(rec["spi"])
        rec["cpi"]            = float(rec["cpi"])

    cur2 = conn.cursor()
    cur2.executemany(upsert_sql, records)
    conn.commit()
    print(f"  Wrote {len(records)} recommendation rows")

    # ── Print top 10 by combined score ─────────────────────────────────────────
    top = df.nlargest(10, "combined_score")[
        ["project_name", "return_score", "risk_score",
         "is_frontier", "nash_preferred", "recommendation", "combined_score"]
    ]
    print("\n── Top 10 projects by combined score ─────────────────────────────")
    print(f"{'Project':<35} {'Ret':>5} {'Risk':>5} {'Frt':>4} {'Nash':>5} {'Tier':<9} {'Score':>6}")
    print("─" * 75)
    for _, r in top.iterrows():
        print(f"{r['project_name']:<35} {r['return_score']:>5.2f} {r['risk_score']:>5.2f} "
              f"{'Y' if r['is_frontier'] else 'N':>4} {'Y' if r['nash_preferred'] else 'N':>5} "
              f"{r['recommendation']:<9} {r['combined_score']:>6.4f}")

    conn.close()
    print("\nDone.")


if __name__ == "__main__":
    np.random.seed(42)
    main()

File: scripts/compute-portfolio-opt.js
/**
 * Compute Portfolio Optimization
 * Markowitz Efficient Frontier + Nash Equilibrium alignment
 */
require('dotenv').config({ path: '.env.local' });
const { Client } = require('pg');

async function computePortfolioOpt() {
  const client = new Client({ connectionString: process.env.DATABASE_URL });
  await client.connect();
  console.log('Connected to database');

  // Get project data from the view
  const projectQuery = `
    SELECT
      project_id,
      project_code,
      project_name,
      project_status,
      trial_phase,
      nme_name,
      nme_code,
      nme_id,
      therapeutic_area,
      task_completion_pct,
      bac,
      spi,
      cpi,
      ev
    FROM v_project_portfolio_base
  `;

  const { rows: projects } = await client.query(projectQuery);
  console.log(`Found ${projects.length} projects`);

  // Clear existing recommendations
  await client.query('DELETE FROM portfolio_recommendation');
  console.log('Cleared existing recommendations');

  // Compute return and risk scores
  const scored = projects.map(p => {
    const spi = parseFloat(p.spi) || 1;
    const cpi = parseFloat(p.cpi) || 1;
    const bac = parseFloat(p.bac) || 1;
    const ev = parseFloat(p.ev) || 0;

    // Return = EV / BAC (how much value earned relative to budget)
    const returnScore = Math.min(ev / bac, 1);

    // Risk = 0.5 * (1 - SPI) + 0.5 * (1 - CPI)
    const riskScore = Math.max(0, Math.min(1, 0.5 * (1 - spi) + 0.5 * (1 - cpi)));

    return {
      ...p,
      return_score: returnScore,
      risk_score: riskScore,
      spi,
      cpi,
    };
  });

  // Determine efficient frontier (Pareto-optimal set)
  // A project is on the frontier if no other project has both higher return AND lower risk
  for (const p of scored) {
    p.is_frontier = !scored.some(other =>
      other.project_id !== p.project_id &&
      other.return_score >= p.return_score &&
      other.risk_score <= p.risk_score &&
      (other.return_score > p.return_score || other.risk_score < p.risk_score)
    );
  }

  // Get task completion by role for Nash equilibrium
  const rolePayoffQuery = `
    SELECT
      p.id AS project_id,
      s.role,
      COUNT(t.id) AS total_tasks,
      COUNT(t.id) FILTER (WHERE t.status = 'DONE') AS done_tasks
    FROM "Project" p
    JOIN "Milestone" m ON m."projectId" = p.id
    JOIN "Task" t ON t."milestoneId" = m.id
    LEFT JOIN "Staff" s ON s.id = t."assigneeId"
    WHERE s.role IN ('PRINCIPAL_SCIENTIST', 'MEDICAL_MONITOR', 'REGULATORY_AFFAIRS')
    GROUP BY p.id, s.role
  `;

  const { rows: roleData } = await client.query(rolePayoffQuery);

  // Build payoff map per project per role
  const payoffMap = {};
  for (const row of roleData) {
    if (!payoffMap[row.project_id]) {
      payoffMap[row.project_id] = {};
    }
    const total = parseInt(row.total_tasks) || 1;
    const done = parseInt(row.done_tasks);
    payoffMap[row.project_id][row.role] = (done / total) * 100;
  }

  // Compute mean payoffs across all projects
  const allPayoffs = { PS: [], MM: [], RA: [] };
  const roleMap = {
    PRINCIPAL_SCIENTIST: 'PS',
    MEDICAL_MONITOR: 'MM',
    REGULATORY_AFFAIRS: 'RA',
  };

  for (const projectId of Object.keys(payoffMap)) {
    const pf = payoffMap[projectId];
    if (pf.PRINCIPAL_SCIENTIST !== undefined) allPayoffs.PS.push(pf.PRINCIPAL_SCIENTIST);
    if (pf.MEDICAL_MONITOR !== undefined) allPayoffs.MM.push(pf.MEDICAL_MONITOR);
    if (pf.REGULATORY_AFFAIRS !== undefined) allPayoffs.RA.push(pf.REGULATORY_AFFAIRS);
  }

  const meanPayoffs = {
    PS: allPayoffs.PS.length ? allPayoffs.PS.reduce((a, b) => a + b, 0) / allPayoffs.PS.length : 50,
    MM: allPayoffs.MM.length ? allPayoffs.MM.reduce((a, b) => a + b, 0) / allPayoffs.MM.length : 50,
    RA: allPayoffs.RA.length ? allPayoffs.RA.reduce((a, b) => a + b, 0) / allPayoffs.RA.length : 50,
  };

  console.log('Mean payoffs:', meanPayoffs);

  // Assign payoffs and Nash preference
  for (const p of scored) {
    const pf = payoffMap[p.project_id] || {};
    p.ps_payoff = pf.PRINCIPAL_SCIENTIST ?? 50;
    p.mm_payoff = pf.MEDICAL_MONITOR ?? 50;
    p.ra_payoff = pf.REGULATORY_AFFAIRS ?? 50;

    // Nash preferred: all roles >= their mean
    p.nash_preferred =
      p.ps_payoff >= meanPayoffs.PS &&
      p.mm_payoff >= meanPayoffs.MM &&
      p.ra_payoff >= meanPayoffs.RA;
  }

  // Compute combined score and recommendation tier
  // Score = 40% Return + 40% (1-Risk) + 20% avg Payoff
  for (const p of scored) {
    const avgPayoff = (p.ps_payoff + p.mm_payoff + p.ra_payoff) / 300; // Normalize to 0-1
    p.combined_score = 0.4 * p.return_score + 0.4 * (1 - p.risk_score) + 0.2 * avgPayoff;

    // Recommendation tier
    if (p.is_frontier && p.nash_preferred) {
      p.recommendation = 'SELECT';
    } else if (p.is_frontier || p.nash_preferred) {
      p.recommendation = 'CONSIDER';
    } else if (p.combined_score >= 0.5) {
      p.recommendation = 'MONITOR';
    } else {
      p.recommendation = 'DEFER';
    }
  }

  // Insert recommendations
  for (const p of scored) {
    await client.query(`
      INSERT INTO portfolio_recommendation (
        project_id, return_score, risk_score, spi, cpi,
        is_frontier, nash_preferred,
        ps_payoff, mm_payoff, ra_payoff,
        recommendation, combined_score, computed_at
      ) VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, NOW())
    `, [
      p.project_id,
      p.return_score.toFixed(6),
      p.risk_score.toFixed(6),
      p.spi.toFixed(6),
      p.cpi.toFixed(6),
      p.is_frontier,
      p.nash_preferred,
      p.ps_payoff.toFixed(4),
      p.mm_payoff.toFixed(4),
      p.ra_payoff.toFixed(4),
      p.recommendation,
      p.combined_score.toFixed(6),
    ]);
  }

  const tierCounts = scored.reduce((acc, p) => {
    acc[p.recommendation] = (acc[p.recommendation] || 0) + 1;
    return acc;
  }, {});

  console.log(`✓ Inserted ${scored.length} portfolio recommendations`);
  console.log('Tier distribution:', tierCounts);
  await client.end();
  console.log('Done');
}

computePortfolioOpt().catch(console.error);
